### ISSUE TITLE
clj-config should validate environment vars

### ISSUE THREAD

**Issue by briprowe** [rentpath/clj-config#15]
Right now clj-config simply checks for existence of declared environment vars. This is nice, but could be better. For example, if an environment variable is expected to be a URL clj-config should be able to validate that it's value is a valid URL.

This will require 2 major changes:

1. We'll have to modify the `defconfig` macro to enable the user to specify a validator.
2. We'll have to modify `clj-config.core/init!` to actually use the provided validators.

**Comment by briprowe** [comment]
I think a good way to move forward would be to create a validation protocol and allow users of this code to extend their types to it. This would enable plain functions and prismatic Schemas to function as validators equally easily.

My main question is how should we modify the `defconfig` macro. Right now we pair symbols to environment variables using maps. If we add validators, a simple key-value pair is no longer enough.

So, I'm thinking we can change `defmacro` to look like one of the following:

```clojure
(defmacro
  :env {sendgrid-url ["SENDGRID_URL" schema/URL]})
```

```clojure
(defmacro
  :env [[sendgrid-url "SENDGRID_URL" schema/URL]])
```

This choice seems arbitrary to me, but can someone see any reason to choose one representation over the other? Are there any other representations that might be better?

**Comment by briprowe** [comment]
Do we want to make this a breaking change, or should we keep the `defconfig` macro backward compatible?

**Comment by briprowe** [comment]
@err ping

**Comment by briprowe** [comment]
We could do this:

```clojure
(defconfig
  :env {^{:validator schema/URL} sendgrid-url "SENDGRID_URL"}
```

This would enable us to add validator info and maintain backwards compatibility. Though, I think it's somewhat ugly.

**Comment by colinrymer** [comment]
Don't worry about breaking backwards compatibility. If it gives us a better API, it's worth it. That's what semver is for.

**Comment by briprowe** [comment]
Cool

**Comment by briprowe** [comment]
After discussing this with some engineers here at FL, we think that `defconfig` might best look like this:

```clojure
(defconfig
  :env [[rex-service-port "REX_SERVICE_PORT" {:restrict schema/Int
                                              :default 8080}]
        [rex-service-protocol "https"]])
```

This way we have an optional map that enables us to specify an arbitrary set of options.

Any thoughts?

CC: @rickhall2000 @err @semperos @halcyon @Christopher-Bui (anyone else I should ping?)

**Comment by briprowe** [comment]
The nested vectors is also a bit less visually noisy than maps with vectors in them.

**Comment by semperos** [comment]
My  $0.02 on this topic:

I'm not in favor of the overall approach of this library, in particular the use of `defconfig` and the `config/init!` check. The use of `defconfig` encourages reliance on context in every namespace that uses it, instead of creating pure functions that accept their config as arguments. In my experience, it's not just a matter of pedantry to maintain referential transparency on this front, it makes for easier testing, maintenance, and refactoring. Also, if the call to `config/init!` is called before all the namespaces that have `defconfig` forms have been loaded (whether by negligence, ignorance of dynamically-loaded namespaces, etc.), the check is futile anyway.

I think this library should be drastically simplified to load Clojure/EDN configuration from conventional places, pull in environment variables based on the .env* files, and pass all of that config back as a Clojure map. If we feel strongly that at configuration-load time we want to ensure certain values are there (as opposed to, for example, including such a check when the app system is bootstrapped, at which point one could have a better idea of what config values are strictly needed), I could see accepting an optional seq of required configuration entries. For anything more complicated, I think returning the Clojure map to the end user provides maximal simplicity and flexibility, allowing them to validate it as needed for their app.

**Comment by briprowe** [comment]
Can you elaborate on your vision for ensuring the existence of config values? Specifically, I'm not understanding where the seq would come from?

**Comment by err** [comment]
I'm indebted to `defconfig`'s compliance check   automagic

**Comment by briprowe** [comment]
Yeah, the def'd vars and lack of writing pure functions are orthogonal. 

Ideally, our components deref the vars when they start, and pass that data around from there.

Is there a way to "componentize" the initialization of config system? Would that allay concerns about when the `config/init!` function runs?

**Comment by briprowe** [comment]
I wish I didn't post the above. 

I'd like to remain focused on the goal of validating configuration at process start-time. We can chat on referential transparency in our next issue :).

**Comment by err** [comment]
I personally am happy with the proposed modification to the `defconfig` macro and the introduction of a validation protocol.

**Comment by semperos** [comment]
@briprowe I appreciate you engaging my comment, and completely understood that it's a bit off-topic :no_entry: I included it here with the goal of limiting how many extra knobs are added to the existing `defconfig`. I'll open a separate issue for further discussion.

If you're going to add validation to `defconfig`, you might consider following suit with how [tools.cli](https://github.com/clojure/tools.cli) handles options (specifically the keyword parts, not the short/long option names):

```clj
(def cli-options
  ;; An option with a required argument
  [["-p" "--port PORT" "Port number"
    :default 80
    :parse-fn #(Integer/parseInt %)
    :validate [#(< 0 % 0x10000) "Must be a number between 0 and 65536"]]
   ;; A non-idempotent option
   ["-v" nil "Verbosity level"
    :id :verbosity
    :default 0
    :assoc-fn (fn [m k _] (update-in m [k] inc))]
   ;; A boolean option defaulting to nil
   ["-h" "--help"]])
```

See [Option Argument Validation](https://github.com/clojure/tools.cli#option-argument-validation) for the semantics here. The syntax here for `defconfig` could still be vector based overall, with either the optional map (as you wrote @briprowe) or with optional key-value entries in a varargs style (as tools.cli does). I'd definitely stick with one of the two (as opposed to just accepting either a schema or a fn), so you can have additional optional config entries like `:default` or `:parse-fn` for example.

**Comment by briprowe** [comment]
@semperos Oh, I didn't mean that the original post was off-topic! I am def interested to hear your thoughts on how we can validate configuration at start-time. I'm especially interested to hear your thoughts on how to generate the optional seq you mentioned.

**Comment by briprowe** [comment]
Here's a PR that implements this: https://github.com/FarmLogs/clj-config/pull/2

I'll modify the README soon.

**Comment by briprowe** [comment]
Just waiting on #14 to be merged to open a PR with the validation code.
